library/Faktory/Worker.hs

```diff
@@ -86,87 +89,82 @@ untilM_ predicate action = do
         untilM_ predicate action
     )
 
--- | Creates a new faktory worker, @'action'@ is ran with @'WorkerConfig'@ before
--- polling begins. Jobs received are passed to @'handler'@. The worker's
--- connection is closed when job processing ends.
-withRunWorker ::
-  (HasCallStack, FromJSON args)
-  =>  Settings
+-- | Returns the @'WorkerId' associated with a @'Worker'.
+workerID :: Worker -> WorkerId
+workerID Worker{workerId} = workerId
+
+-- | Forks a new faktory worker, continuously polls the faktory server for
+-- jobs which are passed to @'handler'@. The client is closed when the forked
+-- thread ends.
+startWorker
+  :: (HasCallStack, FromJSON args)
+  => Settings
   -> WorkerSettings
-  -> (WorkerConfig -> IO a)
   -> (Job args -> IO ())
-  -> IO ()
-withRunWorker settings workerSettings action handler =
-  configureWorker settings workerSettings
-    $ \config -> do
-        void $ action config
-        runWorkerWithConfig handler config
+  -> IO (ThreadId, Worker)
+startWorker settings workerSettings handler = do
+  workerId <- maybe randomWorkerId pure $ settingsId workerSettings
+  isQuieted <- newTVarIO False
+  client <- newClient settings $ Just workerId
+  isDone <- newEmptyMVar
+  let worker = Worker{client, isDone, isQuieted, settings, workerId, workerSettings}
+  tid <-
+    forkFinally
+      ( do
+          beatThreadId <- forkIOWithThrowToParent $ forever $ heartBeat worker
+          finally
+            ( flip runReaderT worker . runWorkerM $
+                untilM_ shouldStopWorker (processorLoop handler)
+                  `catch` (\(_ex :: WorkerHalt) -> pure ())
+            )
+            $ killThread beatThreadId
+      )
+      ( \e -> do
+          closeClient client
+          case e of
+            Left err -> throwIO err
+            Right _ -> putMVar isDone ()
+      )
+  pure (tid, worker)
```

Comment:
Can you put the thread ID in the `Worker` instead? Then add a `workerThreadId :: Worker -> ThreadId`.

Reply:
`Worker` is instantiated before the thread is created and used within the forked thread - is there a way to add it into `Worker` after?

Reply:
Oh, that _is_ annoying. Good catch.

There's a tedious way and a hacky way.

1. Tedious way: create 2 structures, one that is used internally that doesn't have the thread ID, and one exposed internally that doesn't. You'd probably call the internal one `WorkerConfig` or something.
2. Hacky way: add a thread ID field, and set it to be `undefined` until the thread ID is available. This is hacky because if you accidentally evaluate an `undefined` value (e.g. by accessing the field), the entire program crashes.

I think I'd be okay with either approach right now.

Reply:
Hmm, @periodic do you have any better ideas?

Reply:
This is a tough one because you can't reference the tid before it's created and this is monadic and not simple recursion so there's no [tying the knot](https://wiki.haskell.org/Tying_the_Knot).  Also, this value is in a reader, so we can't really "update" it after in a clean way.  I could see passing in an `error` (with a message as opposed to undefined) because it won't be referenced in the finally block.

Is `shouldStopWorker` used anywhere else?  If not, maybe the answer is to inline that one function to not run in `Worker` so that the stuff we do before the worker is started is not run in `Worker`.  Technically the heartbeat isn't a worker, right?

I can't remember if there was a reason not to do that.... maybe it was used elsewhere.

Reply:
OK - I think this would work:
```
data WorkerConfig = WorkerConfig
  { client :: Client
  , settings :: Settings
  , workerId :: WorkerId
  , workerSettings :: WorkerSettings
  }

data Worker = Worker
  { config :: WorkerConfig
  , isQuieted :: TVar Bool
  , isDone :: MVar ()
  , tid :: ThreadId
  }
```
`Worker` hold states information and `WorkerConfig` mostly config information.

then we inline the check for `isQuieted` - eventually we'll have change the heartBeat take the `isQuieted` parameter as well but that's simple enough.

Reply:
Yes, this is great! I like this a lot, much less tedious than I was originally thinking.

Reply:
the annoying part will be getting `workerId` out of worker - does it make sense to have a selector instead of exposing it?